### ISSUE TITLE
override command

### ISSUE THREAD

**Issue by dariooddenino** [justinwoo/spacchetti-cli#25]
We should be able to be able to override packages from the cli.

If having it in a sane way is too complex, we should at least allow to override the package version (which I think already covers 99% of the cases)

**Comment by f-f** [comment]
Do you mean something like a mix of the `extra-deps` and `packages` keys in `stack.yaml`?

Now that we have a `spacchetti.dhall` config, we could add a `override` key in there, in which you could specify a list of git repos that provide some dependencies

**Comment by f-f** [comment]
I'm not sure that would be simpler once you want to change something else than the `version` (e.g. the `repo`). Then we'd have to introduce the syntax `spacchetti override heterogeneous version v0.3.0` to disambiguate the different keys you might want to update. So I don't think the `cli` approach scales.

Instead I'd propose that we add a `overrides` key to the new `spacchetti.dhall`. So let's say I want to override the `console` version in the example config we provide; it would look like this:
```haskell
{-
Welcome to a Spacchetti project!
You can edit this file as you like.
-}
{ name =
    "my-project"
, dependencies =
    [ "effect", "console" ]
, overrides = { console = { version = "v4.0.0" } }  -- new `overrides` key
, packages =
    ./packages.dhall
}
```

This scales because:
- you're not editing the huge `packages` file
- you can add new packages and keys you want to override easily
- no command line implementation required from us

**Comment by justinwoo** [comment]
Overrides should go in the packages.dhall file, that's why it's there locally

**Comment by justinwoo** [comment]
Otherwise the whole raisin deets of using Dhall goes away

**Comment by f-f** [comment]
@justinwoo not necessarily, we can merge the override with Dhall when processing the config. The only difference is that the user doesn't have to manually manage the `packages.dhall`.

But some amount of the usefulness of `packages.dhall` goes away with this change: I think it's ok to not have the `packages.dhall` and generate it from the `spacchetti.dhall` (dhall is now fast enough for that, and we could eventually cache a `packages.dhall` in the `.spacchetti` folder)

**Comment by f-f** [comment]
My vision in the end is to separate the experience of "producing a `packages.dhall`" from "consuming a `packages.dhall`"

**Comment by justinwoo** [comment]
The whole point IS to manage your packages.dhall in whatever way you want using the whole Dhall language, so that you might even have some base overrides in a repo or gist to use in addition to the "remote" in Spacchetti.

**Comment by justinwoo** [comment]
Adding overrides separately on top driven by the CLI honestly is just some hacky shit I don't want to deal with

**Comment by justinwoo** [comment]
Maybe there's some miscommunication here, I hope nobody here wants to add "overrides" to the local project configuration file

**Comment by justinwoo** [comment]
You could opt not to have the packages.dhall file like you said though, if the base is just to provide the remote reference by default and then have the user actually use the Dhall language to modify the set in any way they want

**Comment by f-f** [comment]
@justinwoo I'm still in the process of thinking of a decent user experience, and I think here we should have an open discussion on whatever possibilities we might have to achieve that.
So, whatever idea gets proposed here "it's OK", we evaluate it and see if it makes sense :)

I think you have some good points:
- using the CLI to manipulate Dhall is not the best experience (for both users and maintainers)
- you're right and the `overrides` key in the `spacchetti.dhall` in addition to having the `packages.dhall` doesn't make any sense, let's just integrate the `packages.dhall` into the `spacchetti.dhall`, so one has only one config to manage

**Comment by f-f** [comment]
@justinwoo yeah this sounds good to me. Let me think a bit more about it though

**Comment by justinwoo** [comment]
I think an out of scope question would also be "can someone make me a Dhall-Query program where I can run some queries to get results/modify Dhall sources, even if it's a wrapper around some existing Dhall commands?"

**Comment by justinwoo** [comment]
`dq '.packages | []' spacchetti.dhall`

**Comment by f-f** [comment]
@justinwoo what's the difference with running
```haskell
dhall <<< "(./packages.dhall).packages"
```

**Comment by justinwoo** [comment]
left out the `keys[]`/`values[]` part, along with some other things like `| length`
